lib/vm/src/instance/mod.rs

```diff
@@ -776,6 +782,205 @@ impl Instance {
+        //}
+
+        let ret = unsafe { memory32_atomic_check32(&memory, dst, val) };
+
+        if let Ok(mut ret) = ret {
+            if ret == 0 {
+                ret = self.do_wait(memory_index.as_u32(), dst, timeout);
+            }
+            Ok(ret)
+        } else {
+            ret
+        }
+    }
+
+    /// Perform an Atomic.Wait32
+    pub(crate) fn imported_memory_wait32(
+        &mut self,
+        memory_index: MemoryIndex,
+        dst: u32,
+        val: u32,
+        timeout: i64,
+    ) -> Result<u32, Trap> {
+        let import = self.imported_memory(memory_index);
+        let memory = unsafe { import.definition.as_ref() };
+        //if ! memory.shared {
+        // We should trap according to spec, but official test rely on not trapping...
+        //}
+
+        let ret = unsafe { memory32_atomic_check32(memory, dst, val) };
+
+        if let Ok(mut ret) = ret {
+            if ret == 0 {
+                ret = self.do_wait(memory_index.as_u32(), dst, timeout);
+            }
+            Ok(ret)
+        } else {
+            ret
+        }
+    }
+
+    /// Perform an Atomic.Wait64
+    pub(crate) fn local_memory_wait64(
+        &mut self,
+        memory_index: LocalMemoryIndex,
+        dst: u32,
+        val: u64,
+        timeout: i64,
+    ) -> Result<u32, Trap> {
+        let memory = self.memory(memory_index);
+        //if ! memory.shared {
+        // We should trap according to spec, but official test rely on not trapping...
+        //}
+
+        let ret = unsafe { memory32_atomic_check64(&memory, dst, val) };
+
+        if let Ok(mut ret) = ret {
+            if ret == 0 {
+                ret = self.do_wait(memory_index.as_u32(), dst, timeout);
+            }
+            Ok(ret)
+        } else {
+            ret
+        }
+    }
+
+    /// Perform an Atomic.Wait64
+    pub(crate) fn imported_memory_wait64(
+        &mut self,
+        memory_index: MemoryIndex,
+        dst: u32,
+        val: u64,
+        timeout: i64,
+    ) -> Result<u32, Trap> {
+        let import = self.imported_memory(memory_index);
+        let memory = unsafe { import.definition.as_ref() };
+        //if ! memory.shared {
+        // We should trap according to spec, but official test rely on not trapping...
+        //}
+
+        let ret = unsafe { memory32_atomic_check64(memory, dst, val) };
+
+        if let Ok(mut ret) = ret {
+            if ret == 0 {
+                ret = self.do_wait(memory_index.as_u32(), dst, timeout);
+            }
+            Ok(ret)
+        } else {
+            ret
+        }
+    }
+
+    /// Perform an Atomic.Notify
+    pub(crate) fn local_memory_notify(
+        &mut self,
+        memory_index: LocalMemoryIndex,
+        dst: u32,
+        count: u32,
+    ) -> Result<u32, Trap> {
+        //let memory = self.memory(memory_index);
+        //if ! memory.shared {
+        // We should trap according to spec, but official test rely on not trapping...
+        //}
+
```

Comment:
This seems to be missing an alignment check.

(Spec says to trap on non-aligned notify)

Or is that happening earlier somewhere?

Reply:
I thought it was done earlier, but I need to check as  it's a per compiler thing.